Applications/Resnet/jni/main.cpp

```diff
@@ -65,23 +68,140 @@ static std::string withKey(const std::string &key,
 std::vector<LayerHandle> resnetBlock(const std::string &block_name,
                                      const std::string &input_name, int filters,
                                      int kernel_size, bool downsample) {
-  return {};
+  using ml::train::createLayer;
+
+  auto scoped_name = [&block_name](const std::string &layer_name) {
+    return block_name + "/" + layer_name;
+  };
+  auto with_name = [&scoped_name](const std::string &layer_name) {
+    return withKey("name", scoped_name(layer_name));
+  };
+
+  auto create_conv =
+    [&with_name, filters](const std::string &name, int kernel_size, int stride,
+                          int padding, const std::string &input_layer) {
+      std::vector<std::string> props{
+        with_name(name),
+        withKey("stride", {stride, stride}),
+        withKey("filters", filters),
+        withKey("kernel_size", {kernel_size, kernel_size}),
+        withKey("padding", {padding, padding}),
+        withKey("input_layers", input_layer)};
+
+      return createLayer("conv2d", props);
+    };
+
+  auto create_batch_relu = [&with_name](const std::string &name) {
+    return createLayer("batch_normalization",
+                       {with_name(name), "activation=relu"});
+  };
+
+  /** residual path */
+  LayerHandle a1 = create_conv("a1", 3, downsample ? 2 : 1, 1, input_name);
+  LayerHandle a2 = create_batch_relu("a2");
+  LayerHandle a3 = create_conv("a3", 3, 1, 1, scoped_name("a2"));
+
+  /** skip path */
+  LayerHandle b1 = nullptr;
+  if (downsample) {
+    b1 = create_conv("b1", 1, 2, 0, input_name);
+  }
+
+  const std::string skip_name = b1 ? scoped_name("b1") : input_name;
+
+  LayerHandle c1 = createLayer(
+    "Addition",
+    {with_name("c1"), withKey("input_layers", {scoped_name("a3"), skip_name})});
+
+  LayerHandle c2 = create_batch_relu(""); /// use block_name itself.
+
+  if (downsample) {
+    return {a1, a2, a3, b1, c1, c2};
+  } else {
+    return {a1, a2, a3, c1, c2};
+  }
 }
 
-/**s
+/**
  * @brief Create resnet 18
  *
- * @return ModelHandle to create model
+ * @return vector of layers that contain full graph of resnet18
  */
-ModelHandle createResnet18() {
+std::vector<LayerHandle> createResnet18Graph() {
+  using ml::train::createLayer;
+
   std::vector<LayerHandle> layers;
-  return nullptr;
+
+  layers.push_back(
```

Comment:
i wonder where is the input layer? :)

Reply:
when we have input_shape property, we don't need input layer. Could you clarify this?

Reply:
Isn't input layer necessary for model creation. I maybe wrong here.

Reply:
afaik, if input_shape is there it is treated input. I have tested it's running fine but if there is a problem, I'll look into it :)

Reply:
cool.